### ISSUE TITLE
Hook paket into dotnet add

### ISSUE THREAD

**Issue by forki** [dotnet/cli#8218]
Is there a way to let other tools hook into dotnet add?
We would like to call `paket add` from there

/cc @enricosada

**Comment by livarcocc** [comment]
That's not possible and so far we have no plans of adding such support.

**Comment by forki** [comment]
May I ask why not? Most other commands are pluggable

**Comment by livarcocc** [comment]
If you are referring to build/run/publish/pack/test etc, they run as MSBuild commands and therefore they benefit from the extensibility points of MSBuild itself, with props/targets.

`dotnet add` is built in into the CLI. They use the MSBuild project construction APIs to manipulate the project files, or nuget APIs to add package references and these APIs do not have the same extensibility as a MSBuild build.

**Comment by forki** [comment]
Yes but in general. The dotnet cli was always advertised as "driver" for
other commands and as approach for unification. In that spirit I think it
should work for add as well. Currently dotnet add just breaks projects that
use paket. Which is not surprising, but it's not really in the spirit of
the dotnet cli. At least not in the sense I understood it

Am 12.12.2017 20:30 schrieb "Livar" <notifications@github.com>:

**Comment by livarcocc** [comment]
It has a set of commands in which it adds as a drivers, like the commands that eventually invoke MSBuild and I listed above.

It also has a set of commands for project modification, like add, remove, list, sln commands. These commands have no counterpart in other parts of the stack, i.e MSBuild and as such are implemented in the CLI directly.

How does `dotnet add` breaks paket? It does so for project references and package references?

**Comment by livarcocc** [comment]
Ok, I guess I understand now.

In this case, I would suggest looking at global tools and making paket a global tool. Then you could do `dotnet install tool paket` and then people could simply run paket. If you make paket a portable application, this should be very straight forward and we can help you making it a tool.

**Comment by forki** [comment]
See we are looking into creating a global tool. That's not my concern. The
concern is that you have a command that adds packages to a project, but it
does that not in the way we need it to do. And other tools start to  work
on top of the dotnet cli. So these things will also not do the right thing.
But if you make add and remove pluggable, then everything could work
automatically in "correct" way.

Am 12.12.2017 22:32 schrieb "Livar" <notifications@github.com>:

**Comment by livarcocc** [comment]
I understand you concern and your point. However, this is the very first ask we have had for such extensibility model for dotnet add and as such we don't have a design for it nor a timeline for when we would be able to fund implementing such extensibility.

Perhaps, if you can write up a design doc for this and share with the team and once it is reviewed and accepted, make some PRs, this is something that we would gladly accept.

**Comment by forki** [comment]
Lol of course it's the first request. Since paket is the only other package
manager around.

Am 14.12.2017 21:07 schrieb "Livar" <notifications@github.com>:

**Comment by Jaygem** [comment]
Is this request relevant as of now?
It could be useful to add paket as a dotnet tool.
